ethcore/src/miner/miner.rs

```diff
@@ -799,21 +802,34 @@ impl miner::MinerService for Miner {
 
 		trace!(target: "own_tx", "Importing transaction: {:?}", pending);
 
-		let client = self.pool_client(chain);
-		let imported = self.transaction_queue.import(
-			client,
-			vec![pool::verifier::Transaction::Local(pending)]
-		).pop().expect("one result returned per added transaction; one added => one result; qed");
+		// treat the tx as local if the option is enabled, or if we have the account
+		let sender = pending.sender();
+		let treat_as_local = self.options.tx_queue_allow_unknown_local
+			|| self.accounts.as_ref().map(|accts| accts.has_account(sender)).unwrap_or(false);
 
-		// --------------------------------------------------------------------------
-		// | NOTE Code below requires sealing locks.                                |
-		// | Make sure to release the locks before calling that method.             |
-		// --------------------------------------------------------------------------
-		if imported.is_ok() && self.options.reseal_on_own_tx && self.sealing.lock().reseal_allowed() {
-			self.prepare_and_update_sealing(chain);
-		}
+		if treat_as_local {
+			let client = self.pool_client(chain);
+			let imported = self.transaction_queue.import(
+				client,
+				vec![pool::verifier::Transaction::Local(pending)]
+			).pop().expect("one result returned per added transaction; one added => one result; qed");
 
-		imported
+            // --------------------------------------------------------------------------
+            // | NOTE Code below requires sealing locks.                                |
+            // | Make sure to release the locks before calling that method.             |
+            // --------------------------------------------------------------------------
+            if imported.is_ok() && self.options.reseal_on_own_tx && self.sealing.lock().reseal_allowed() {
+                self.prepare_and_update_sealing(chain);
+            }
+
+
+			imported
+		} else {
```

Comment:
I would suggest we:

* Leave `import_own_transaction` unchanged.
* Create another function, probably called `import_local_or_external_transaction` (or maybe a better name, can't think of right now), which dispatch to `import_own_transaction` or `import _external_transactions`.
* Make `eth_submitRawTransaction` call that function instead, through `Dispatcher`.

The issue is that `import_own_transaction` is used in many other places in `ethcore`, and this change may break some of the previous assumptions.

Reply:
Yup, agreed. Much more elegant . Will get on that tomorrow.

Reply:
How does `import_foreign_transaction` sound? Or maybe just `import_unknown_local_transaction`, which corresponds to the argument name very well. 

On the one hand introducing a new term is not good, but coming up with a good name is hard. Maybe `normalise_unknown_local`? like mixing them in with external instead of treating like known locals. I think the External/Local thing makes sense and should be left how it is (no redefining what a local tx is). (though maybe better names would be Indirect/Direct transactions (either that  you get them second-hand or directly.)

Reply:
Both sound fine to me. :)